R/object_overwrite_linter.R

```diff
@@ -0,0 +1,94 @@
+#' Block assigning any variables whose name clashes with a `base` R function
+#'
+#' Re-using existing names creates a risk of subtle error best avoided.
+#'   Avoiding this practice also encourages using better, more descriptive names.
+#'
+#' @param packages Character vector of packages to search for names that should
+#'   be avoided. Defaults to the most common default packages: base, stats,
+#'   utils, tools, methods, graphics, and grDevices.
+#'
+#' @examples
+#' # will produce lints
+#' code <- "function(x) {\n  data <- x\n  data\n}"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = object_overwrite_linter()
+#' )
+#'
+#' code <- "function(x) {\n  lint <- 'fun'\n  lint\n}"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = object_overwrite_linter(packages = "lintr")
+#' )
+#'
+#' # okay
+#' code <- "function(x) {\n  data('mtcars')\n}"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = object_overwrite_linter()
+#' )
+#'
+#' code <- "function(x) {\n  data <- x\n  data\n}"
+#' writeLines(code)
+#' lint(
+#'   text = code,
+#'   linters = object_overwrite_linter(packages = "base")
+#' )
+#'
+#' @evalRd rd_tags("object_overwrite_linter")
+#' @seealso
+#'  - [linters] for a complete list of linters available in lintr.
+#'  - <https://style.tidyverse.org/syntax.html#object-names>
+#' @export
+object_overwrite_linter <- function(
+    packages = c("base", "stats", "utils", "tools", "methods", "graphics", "grDevices")) {
+  for (package in packages) {
+    if (!requireNamespace(package, quietly = TRUE)) {
+      stop("Package '", package, "' is not available.")
+    }
+  }
+  pkg_exports <- lapply(packages, getNamespaceExports)
+  pkg_exports <- data.frame(
+    package = rep(packages, lengths(pkg_exports)),
+    name = unlist(pkg_exports),
+    stringsAsFactors = FALSE
+  )
+
+  # test that the symbol doesn't match an argument name in the function
```

Comment:
This should be in the examples to show the feature.

Reply:
I left it out since I thought we might drop this exclusion. But I remember why we put it in to begin with -- changing function signatures could mean backwards-incompatible changes, so renaming an argument that's overwriting an object is _much_ harder than naming an internal object. I guess we could check if the function is exported but even then it's hard to tell if this object is in a script that's `source()`'d by another script, or maybe this is working under {box}, etc.

Reply:
I also think the argument case is less bad in general because of the locality.

Reply:
Added to `?object_overwrite_linter` examples